compiler/rustc_const_eval/src/const_eval/eval_queries.rs

```diff
@@ -132,8 +132,8 @@ pub(super) fn op_to_const<'tcx>(
         // functionality.)
         _ => false,
     };
-    let immediate = if force_as_immediate {
-        Right(ecx.read_immediate(op).expect("normalization works on validated constants"))
+    let immediate = if force_as_immediate && let Ok(imm) = ecx.read_immediate(op) {
```

Comment:
This is "catching" an interpreter error, and this codepath is used for regular consts, not just for diagnostics. That's fragile, I don't think we should do this.

Reply:
IIUC, this code path is mostly an optimization. Nothing would break if we were to set `force_as_immediate` as false: we do not *need* to represent scalars as a `ConstValue::Scalar`, it's just more efficient.

Reply:
That's wrong, many parts of the compiler rely on `try_to_scalar` to work for integer types, and will ICE if we don't use the "immediate" code path here.

Reply:
There's a comment just a few lines above explaining that:
```
    // All scalar types should be stored as `ConstValue::Scalar`. This is needed to make
    // `ConstValue::try_to_scalar` efficient; we want that to work for *all* constants of scalar
    // type (it's used throughout the compiler and having it work just on literals is not enough)
    // and we want it to be fast (i.e., don't go to an `Allocation` and reconstruct the `Scalar`
    // from its byte-serialized form).
```
Please suggest how we could word that more clearly to avoid such misunderstandings in the future.